Add documentation for the EnvFiles feature gate (#51415)
kubernetes/website: content/en/docs/tasks/inject-data-application/define-environment-variable-via-file.md

During container initialization, the kubelet retrieves environment variables 
from specified files in the `emptyDir` volume and exposes them to the container.

{{< note >}}
All container types (initContainers, regular containers, sidecars containers,
and ephemeral containers) support environment variable loading from files.

While these environment variables can store sensitive information, 
`emptyDir` volumes don't provide the same protection mechanisms as
dedicated Secret objects. Therefore, exposing confidential environment variables 
to containers through this feature is not considered a security best practice.
{{< /note >}}

Create the Pod:

```shell
kubectl apply -f https://k8s.io/examples/pods/inject/envars-file-container.yaml
```

Verify that the container in the Pod is running:

```shell
# If the new Pod isn't yet healthy, rerun this command a few times.
kubectl get pods
```

Check container logs for environment variables:

```shell
kubectl logs dapi-test-pod -c use-envfile | grep DB_ADDRESS
```

The output shows the values of selected environment variables:

```
DB_ADDRESS=address
```

> ## `.env` File Syntax

Reviewer: are comments supported?

---
title: Defining Environment Variable Values via File
content_type: task
weight: 30
---

<!-- overview -->

{{< feature-state feature_gate_name="EnvFiles" >}}

This page show how to configure environment variables for containers in a Pod via file.

> To use this feature, ensure your Kubernetes version v1.34 or later

Reviewer 2: Put this inside "prerequisites". Also specify `min-kubernetes-server-version` in the front matter (see other pages for an example).

You should then use a `version-check` shortcode.

with the `EnvFiles` feature gate enabled.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}}

<!-- steps -->

> ## Using File to Define Environment Variables for Containers in a Pod

Reviewer 2: ```suggestion
## How the design works 
```

In this exercise, you will create a Pod that sources environment variables from files, 
projecting these values into the running container.

{{% code_sample file="pods/inject/dapi-envars-file-container.yaml" %}}

In this manifest, you can see the `initContainer` mounts an `emptyDir` volume and writes environment variables to a file within it,
and the regular containers reference both the file and the environment variable key 
through the `fileKeyRef` field without needing to mount the volume. 
When `optional` field is set to false, the specified `key` in `fileKeyRef` must exist in the environment variables file.

> Please note that the volume will only be mounted to the environment variable producer container 

Reviewer 2: ```suggestion
The volume will only be mounted to the container that writes to the file
```

(`initContainer`), while the consumer container that consumes the environment variable will not have the volume mounted.

> The env file format adheres to the [RFC .env standard](https://smartmob-rfc.readthedocs.io/en/latest/2-dotenv.html#formal-specification).

Reviewer 2: That's an odd link for an "RFC". We could just link to Wikipedia?

Author: Wow, after some investigation, I've discovered this is actually a fascinating misunderstanding.

The specification at https://smartmob-rfc.readthedocs.io/en/latest/2-dotenv.html#formal-specification was not published by IETF. It was created by an organization or individual named "smartmob" on GitHub and hasn't gained significant attention. There are also no formal reliability guarantees associated with this document.

The original repository for this document is: https://github.com/smartmob-project/smartmob-rfc

Although we've implemented our code based on this specification, we should probably remove all references to this document from our KEPs, code comments, and documentation. Instead, we should replace those references with concrete descriptions of our actual implementation behavior. This will prevent unnecessary misunderstandings for users leveraging this feature.

@SergeyKanzhelev WDYT?

Author: @SergeyKanzhelev Pinging again to prevent the message from being overlooked.

Reviewer: let's remove the link.

I think we will need to describe the format ourselves at this point. We can say "generally something that will set env variables by doing:

```
set -a # automatically export all variables
source .env
set +a
```

this will help with allowing comments with `#` and explain how to do newlines.

During container initialization, the kubelet retrieves environment variables 

> from specified files in the `emptyDir` volume and exposes them to the container.

Reviewer 2: Does it have to be a volume of type `emptyDir`? Readers may think that it does.

Author: Yes, we only allow volumes of type `emptyDir`.

{{< note >}}

> All container types (`initContainers`, `regular containers`, `sidecars containers`,

Reviewer 2: Don't use backticks here.

and `ephemeral containers`) support environment variable loading from files.

While these environment variables can store sensitive information, 

> please note that `emptyDir` volumes don't provide the same protection mechanisms as 

Reviewer 2: ```suggestion
`emptyDir` volumes don't provide the same protection mechanisms as
```

dedicated Secret objects. Therefore, exposing confidential environment variables 

> to containers through this feature is not considered a security best practice.

Reviewer 2: I am not sure about this; the Secret API also has problems.

{{< /note >}}

Create the Pod:

```shell

> kubectl apply -f https://k8s.io/examples/pods/inject/dapi-envars-file-container.yaml

Reviewer 2: This is not a downward API example; we should use a different path.

```

Verify that the container in the Pod is running:

```shell
# If the new Pod isn't yet healthy, rerun this command a few times.
kubectl get pods
```

Check container logs for environment variables:

```shell
kubectl logs dapi-test-pod -c use-envfile | grep DB_ADDRESS
```

The output shows the values of selected environment variables:

```
DB_ADDRESS=address
```

## {{% heading "whatsnext" %}}

* Read [Defining Environment Variables for a Container](/docs/tasks/inject-data-application/define-environment-variable-container/)
* Read the [`spec`](/docs/reference/kubernetes-api/workload-resources/pod-v1/#PodSpec)
  API definition for Pod. This includes the definition of Container (part of Pod).
* Read the list of [available fields](/docs/concepts/workloads/pods/downward-api/#available-fields) that you
  can expose using the downward API.

> Read about Pods, containers and environment variables in the legacy API reference:

Reviewer 2: I wouldn't link to that one page reference. Consider linking to the multi page reference instead.
